### ISSUE TITLE
"r10k deploy module" not deploying module

### ISSUE THREAD

**Issue by CCTJNII** [puppetlabs/r10k#329]
I'm seeing behavior where "r10k deploy module" isn't deploying module changes.  An environment deploy does deploy the changes.  "r10k deploy module" does not produce any output, even at debug2 log level.

In the following example I added a line containing 'tmpwatch' to a manifest and am trying to deploy it:

```
[r10k-deploy@6636b4825d26 /]$ grep tmpwatch /etc/puppet/environments/stage/modules/miscpackages/ -r
[r10k-deploy@6636b4825d26 /]$ r10k deploy module miscpackages
[r10k-deploy@6636b4825d26 /]$ r10k deploy module miscpackages -v debug2
[r10k-deploy@6636b4825d26 /]$ grep tmpwatch /etc/puppet/environments/stage/modules/miscpackages/ -r
[r10k-deploy@6636b4825d26 /]$ r10k deploy environment stage -p
[r10k-deploy@6636b4825d26 /]$ grep tmpwatch /etc/puppet/environments/stage/modules/miscpackages/ -r
/etc/puppet/environments/stage/modules/miscpackages/manifests/init.pp:             'tmpwatch', # Removed 2015-02-27: Causes failures in apps that use /tmp
[r10k-deploy@6636b4825d26 /]$ 
```

I do not see anything obviously wrong in my config:

```
[r10k-deploy@6636b4825d26 /]$ grep miscpackages /etc/puppet/environments/stage/Puppetfile 
mod 'miscpackages',
   :git => "git@cweogitlab.sys.comcast.net:puppetmodules/miscpackages.git"
```

I am running r10k 4.4.1:

```
[r10k-deploy@6636b4825d26 /]$ gem list | grep r10k
r10k (1.4.1)

```

This impacts our workflow as our webhook calls 'r10k deploy module' on module updates which is not functioning correctly.  Please advise.

**Comment by adrienthebo** [comment]
What does `r10k deploy display` return? Does `r10k deploy environment -p -v` indicate that it's updating anything? The fact that no commands are running at all might indicate that there's a configuration error.

**Comment by CCTJNII** [comment]
I'm running inside a Docker container to easily wrap the webhook, with the cache inside the container.  I'm noticing that 'r10k deploy display' and 'r10k deploy module' do not appear to update the cache, whereas 'r10k deploy environment' does.  Could this be an issue with the cache not being current?  The webhook container is long lived so the cache there should be complete.

After running 'r10k deploy environment prod -p' display returns my branches:

```
[r10k-deploy@9deb6e3d6342 /]$ r10k deploy display
[R10K::Action::Deploy::Display - ERROR] Command exited with non-zero exit code:
Command: git --git-dir /home/r10k-deploy/r10k_cache/git@cweogitlab.sys.comcast.net-puppetmodules-environments.git for-each-ref refs/heads --format %(refname)
Stderr:
fatal: Not a git repository: '/home/r10k-deploy/r10k_cache/git@cweogitlab.sys.comcast.net-puppetmodules-environments.git'
Exit code: 128
[r10k-deploy@9deb6e3d6342 /]$ r10k deploy environment prod -p
[r10k-deploy@9deb6e3d6342 /]$ r10k deploy display
local (/etc/puppet/environments)
    - master
    - prod
    - stage
    - test
```

'r10k deploy environment -p -v' does deploy as expected, and I do see it updating my modules.  I omitted the -v from my initial post to save pasting the output from all my modules.  Deploying the environment does update the module, and is my current workaround.

**Comment by adrienthebo** [comment]
`r10k deploy display` shouldn't touch the network, so it won't update any cached repos. `r10k deploy module` should update the cache of the module being updated. The fact that you're not seeing any output with `-v debug2` is really interesting; I would assume that you would see something. Let me do some investigation on my end to confirm that you should see some logged output.

**Comment by adrienthebo** [comment]
If you're running `r10k deploy module <something>` with `-v debug2`, you should see a message logged for every single module that wasn't updated; in my test environment there were 7.8k messages logged. If you're not seeing anything, that seems to indicate that r10k thinks that you don't have any environments. If you try to update a differently named module, does the same thing happen?

**Comment by CCTJNII** [comment]
This has to do with the cache.

Starting with a bare cache I have to deploy the environment before module updates will work.  Once the environment has been deployed and the cache updated then 'deploy module' works and prints debug info.  If the cache is invalid then I see the behavior initially described.

Is this by design that it seems to be looking in the cache to know what to update instead of /etc/puppet/environments?  I didn't break into the exact behavior but I did prove that persisting the cache seems to solve this problem, and fits with my failure mode.  I had expected that I could destroy the cache without repercussions and that a module update would pull data from /etc/puppet/environments.

**Comment by CCTJNII** [comment]
I was able to resolve this by ensuring the cache is persistent.  Ultimately I had some confusion as the term 'cache' made me believe that the cache dir could be wiped between runs, I want to suggest renaming this but am at a loss for something better.  (It isn't really spool, lib dir maybe?)

I think this bug is sufficient documentation as this was largely caused by running it in Docker, so I'm going to mark it as solved.  It should probably be re-opened if the cache dir is expected to be nukable, though.

**Comment by adrienthebo** [comment]
Were you removing the cache, and then running `r10k deploy module <modulename>`? If you wipe the cache directory, the git repositories should be restored but you're taking an enormous performance hit by nuking the cache. Even if the cache was fully rebuilt between runs, you would be re-cloning the repos every time, so you do want to make sure the cache persists.

**Comment by CCTJNII** [comment]
The cache was in the container, so new containers would start from no cache but long lived containers would build up a cache.  I had assumed that all commands would work from no cache, so I didn't persist the cache outside the container.  The module update webhook worked if it had previously deployed the environment and as such had built up the cache, but failed if the cache was not yet built.  This would happen after the container was restarted, I ran into this issue this morning as a full environment deploy had not yet been done through the running container so the cache wasn't built yet.

I did find that the repos are not restored into the cache on a module deploy, but are on an environment deploy.  So if 'r10k deploy module' is expected to work with a broken cache then we should reopen this.

**Comment by adrienthebo** [comment]
Trying to do anything with a deleted cache is unsupported; if you directly use git commands inside of any checked out repositories, they'll fail hard because all of their objects are missing. It's by happenstance that the cache will be cleanly rebuilt if deleted and I want to preserve this, but right now it's undefined as to what'll actually happen, especially because git itself will go nuts.

**Comment by CCTJNII** [comment]
So then what is the correct way to init the cache?  Environment checkout?  I don't see cache init covered in the readme or the FAQ.

I think this is behavior that should be defined, even if it is nothing more than a simple check and error, just to ensure r10k behaves long term as environments drift and users do unexpected things.  So if the user tries to do a module update with no cache it would be nice if it could report an error that the required cache is missing.  Is that possible?

**Comment by adrienthebo** [comment]
It's been on my todo list to document this for a while (https://github.com/puppetlabs/r10k/issues/147) but I've had hard time getting around to it. Basically, what's happening is that the working repositories are directly linked to the cached git repository. Those working repositories have no objects of their own and simply reuse the objects in the central repo. If you delete the cached repositories, it's effectively identical to doing `rm -rf .git/objects` which lobotomizes the git repository. Given this, the solution is to not delete the cache because of how badly it breaks git.

By coincidence this problem can usually be avoided due to overzealous cache refreshing, which in the past has served to mask this problem. However, the expectation for `r10k deploy module` is that you've already deployed everything and want to update a select set of modules, in which case there's no reason to update the cache for the main Git repository. In fact, the Git source code handles the case where the cache is empty by treating it as [having no branches](https://github.com/puppetlabs/r10k/blob/master/lib/r10k/source/git.rb#L71-L83). Trying to preemptively fetch this cache would be really ugly and trigger a lot of code at the wrong time, and in normal operation it'll never happen.

So long story short, don't delete the cache - if you're using some sort of containerized environment, make sure the cached repositories are available with the actual working repositories. There's so much undefined behavior across the board that there aren't a lot of good ways to handle this. In the long run there should be better messaging around this, but since it's so far out of the norm that I won't be able to add that in the immediate future.

**Comment by adrienthebo** [comment]
PR for documentation submitted at https://github.com/puppetlabs/r10k/pull/358
